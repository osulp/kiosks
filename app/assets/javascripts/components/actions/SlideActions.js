export const SET_SLIDES = 'SET_SLIDES';
export const REFRESH_SLIDES = 'REFRESH_SLIDES';

export function setSlides(slides) {
  slides.push({id: 100, content: "absdg"});
  console.log(slides);
  return {
    type: SET_SLIDES,
    slides: slides
  };
}

export function refreshSlides(){
  // TODO: Ajax call to query the current set of slides?
}
