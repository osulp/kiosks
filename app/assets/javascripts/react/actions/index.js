export * from './kioskActions';
export * from './touchActions';
