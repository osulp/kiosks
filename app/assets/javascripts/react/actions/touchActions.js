import fetch from 'isomorphic-fetch';
import moment from 'moment';
import {addError, setSlides} from './kioskActions';

export const SET_MAPS = 'SET_MAPS';
export const setMaps = (maps) => {
  return {
    type: SET_MAPS,
    maps
  };
};

export const FETCHING_SLIDES = 'FETCHING_SLIDES';
export const fetchingSlides = () => {
  return {
    type: FETCHING_SLIDES
  };
};

export const FETCHED_SLIDES = 'FETCHED_SLIDES';
export const fetchedSlides = () => {
  return {
    type: FETCHED_SLIDES
  };
};

/**
 * A redux-thunk async action that fetches from the server and handles the return by dispatching a regular
 * action depending on success or error.
 * @param url - the url to the server for fetching slides
 * @returns {Redux Async Action} - see http://redux.js.org/docs/advanced/AsyncActions.html
 */
export const fetchSlides = (url) => {
  return (dispatch) => {
    dispatch(fetchingSlides());
    return fetch(`${url}.json`)
      .then(response => response.json())
      .then(json => {
        dispatch(setSlides(json.slides));
        // it's fast, so let the fetching_slides animation run for 800ms more. ;)
        setTimeout(() => {
          dispatch(fetchedSlides());
        }, 800);
      })
      .catch(err => {
        dispatch(addError({message: err.message, code: err.code}));
        dispatch(fetchedSlides());
      });
  };
};

export const SET_HOURS = 'SET_HOURS';
export const setHours = (hours) => {
  return {
    type: SET_HOURS,
    hours
  };
};

export const FETCHING_HOURS = 'FETCHING_HOURS';
export const fetchingHours = () => {
  return {
    type: FETCHING_HOURS
  };
};

export const FETCHED_HOURS = 'FETCHED_HOURS';
export const fetchedHours = () => {
  return {
    type: FETCHED_HOURS
  };
};

/**
 * A redux-thunk async action that fetches from the server and handles the return by dispatching a regular
 * action depending on success or error.
 * @param url - the url to the server for fetching slides
 * @returns {Redux Async Action} - see http://redux.js.org/docs/advanced/AsyncActions.html
 */
export const fetchHours = (url, dates) => {
  return (dispatch) => {
    dispatch(fetchingHours());
    // Default to this weeks dates if dates is undefined
    if (!dates) {
      dates = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"]
        .map(d => moment().day(d));
    }
    return fetch(url, {
      method: 'POST',
      headers: {
        'Accept': 'application/vnd.kiosks.v1',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({dates:[...dates]})
    })
      .then(response => {
        switch(response.status) {
          case 200:
            return response.json();
          case 404:
            return {"error": "Hours not found for the selected date."};
          default:
            return {"error": "Error fetching hours, please notify the help desk."};
        }
      })
      .then(json => {
        dispatch(setHours(json));
        // it's fast, so let the fetching_hours animation run for 800ms more. ;)
        setTimeout(() => {
          dispatch(fetchedHours());
        }, 800);
      })
      .catch(err => {
        dispatch(addError({message: err.message, code: err.code}));
        dispatch(fetchedHours());
      });
  };
};
