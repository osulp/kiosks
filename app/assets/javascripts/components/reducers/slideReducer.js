import { SET_SLIDES, REFRESH_SLIDES } from '../actions/slideActions';

const slideReducer = (state = {}, action) => {
   switch (action.type) {
    case REFRESH_SLIDES:
      return Object.assign({}, action.slides);
    case SET_SLIDES:
      return action.slides;
    default:
      return state;
  }
}

export default slideReducer;