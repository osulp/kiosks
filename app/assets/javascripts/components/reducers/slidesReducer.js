import { SET_SLIDES, REFRESH_SLIDES } from '../actions/SlideActions';

const slidesReducer = (state = {}, action) => {
   switch (action.type) {
    case REFRESH_SLIDES:
      return Object.assign({}, action.slides);
    case SET_SLIDES:
      return action.slides;
    default:
      return state;
  }
}

export default slidesReducer;