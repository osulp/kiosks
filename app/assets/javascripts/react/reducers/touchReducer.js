import {FETCHED_SLIDES, FETCHING_SLIDES, FETCHED_HOURS, FETCHING_HOURS, SET_HOURS, SET_MAPS} from '../actions/touchActions';

export const initial_state = {
  is_fetching_slides: false,
  is_fetching_hours: false,
  hours: {},
  maps: []
};

const touchReducer = (state = initial_state, action) => {
  switch (action.type) {
    case FETCHED_SLIDES:
      return Object.assign({}, state, { is_fetching_slides: false });
    case FETCHING_SLIDES:
      return Object.assign({}, state, { is_fetching_slides: true });
    case FETCHED_HOURS:
      return Object.assign({}, state, { is_fetching_hours: false });
    case FETCHING_HOURS:
      return Object.assign({}, state, { is_fetching_hours: true });
    case SET_HOURS:
      return Object.assign({}, state, { hours: action.hours });
    case SET_MAPS:
      return Object.assign({}, state, { maps: action.maps });
    default:
      return state;
  }
};

export default touchReducer;