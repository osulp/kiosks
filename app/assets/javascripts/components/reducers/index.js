import { combineReducers } from 'redux';
import slidesReducer from './slidesReducer';

const rootReducer = combineReducers({
  slides: slidesReducer
});

export default rootReducer;