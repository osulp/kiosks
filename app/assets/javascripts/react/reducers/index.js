import { combineReducers } from 'redux';
import kioskReducer from './kioskReducer';
import modalReducer from './modalReducer';
import touchReducer from './touchReducer';
import circReducer from './circReducer';

export const REDUCERS_SHAPE = {
  kiosk: kioskReducer,
  modal: modalReducer,
  touch: touchReducer,
  circ: circReducer
};

export default combineReducers(REDUCERS_SHAPE);
