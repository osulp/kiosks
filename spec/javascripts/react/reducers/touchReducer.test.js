import {default as reducerCreator} from '../../../../app/assets/javascripts/react/reducers/touchReducer';
import * as actionCreator from '../../../../app/assets/javascripts/react/actions/touchActions';
import * as factories from '../.factories';

const empty_state = undefined;

describe('Reducers::Touch', () => {
  it('matches the snapshot', () => {
    expect(reducerCreator).toMatchSnapshot();
  });
  describe('fetchingSlides action', () => {
    it('matches the snapshot', () => {
      expect(reducerCreator(empty_state, actionCreator.fetchingSlides())).toMatchSnapshot();
    })
  });
  describe('fetchingHours action', () => {
    it('matches the snapshot', () => {
      expect(reducerCreator(empty_state, actionCreator.fetchingHours())).toMatchSnapshot();
    })
  });
  describe('fetchedSlides action', () => {
    it('matches the snapshot', () => {
      expect(reducerCreator(empty_state, actionCreator.fetchedSlides())).toMatchSnapshot();
    })
  });
  describe('fetchedHours action', () => {
    it('matches the snapshot', () => {
      expect(reducerCreator(empty_state, actionCreator.fetchedHours())).toMatchSnapshot();
    })
  });
  describe('setHours action', () => {
    it('matches the snapshot', () => {
      expect(reducerCreator(empty_state, actionCreator.setHours(factories.hours))).toMatchSnapshot();
    })
  });
  describe('setMaps action', () => {
    it('matches the snapshot', () => {
      expect(reducerCreator(empty_state, actionCreator.setMaps(factories.map))).toMatchSnapshot();
    })
  });
});
