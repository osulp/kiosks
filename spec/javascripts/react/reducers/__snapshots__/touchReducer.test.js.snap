exports[`Reducers::Touch fetchedHours action matches the snapshot 1`] = `
Object {
  "hours": Object {},
  "is_fetching_hours": false,
  "is_fetching_slides": false,
  "maps": Array [],
}
`;

exports[`Reducers::Touch fetchedSlides action matches the snapshot 1`] = `
Object {
  "hours": Object {},
  "is_fetching_hours": false,
  "is_fetching_slides": false,
  "maps": Array [],
}
`;

exports[`Reducers::Touch fetchingHours action matches the snapshot 1`] = `
Object {
  "hours": Object {},
  "is_fetching_hours": true,
  "is_fetching_slides": false,
  "maps": Array [],
}
`;

exports[`Reducers::Touch fetchingSlides action matches the snapshot 1`] = `
Object {
  "hours": Object {},
  "is_fetching_hours": false,
  "is_fetching_slides": true,
  "maps": Array [],
}
`;

exports[`Reducers::Touch matches the snapshot 1`] = `[Function]`;

exports[`Reducers::Touch setHours action matches the snapshot 1`] = `
Object {
  "hours": Object {
    "2000-01-01": Object {
      "close": "2am",
      "open": "1am",
      "sortable_date": "2000-01-01",
      "string_date": "Jan 01, 2000",
    },
  },
  "is_fetching_hours": false,
  "is_fetching_slides": false,
  "maps": Array [],
}
`;
